### DISCUSSION TITLE
Question about using quantum algorithms from liboqs

### DISCUSSION THREAD

**Discussion by NovakovicA** [open-quantum-safe/openssl#366] [Q&A] [answered]
Hello,

The question might be stupid, in which case I apologise in advance.

Is there a way to use quantum algorithms (NTRU, FrodoKEM etc) within this OpenSSL fork as you would, for instance, use AES, RSA and other algorithms from OpenSSL (non-related to TLS), having both quantum and non-quantum algorithms within the same library? It's my first time looking at the OpenSSL code so I might have missed some documentation files related to this. Any help pointing me out to the right direction/needed documentation is greatly appreciated.

**Comment by baentsch** [comment, marked as answer]
Well,Yes and No: It depends on the algorithm family: The [quantum-safe key exchange algorithms](https://github.com/open-quantum-safe/openssl#key-exchange) can be used as-is exactly only as intended, i.e., as KEX algorithms, not as plain streaming ciphers (like say AES). The [QS signature algorithms](https://github.com/open-quantum-safe/openssl#authentication) in turn work pretty much identical to classic signature algorithms (without the need to use a digest before feeding them, although that still can be done), e.g., see the [CMS example](https://github.com/open-quantum-safe/openssl#cms-demo).

**Reply by NovakovicA** [reply]
Seems like I missed a very important part of my question - I'm looking to use the quantum algorithms (with a focus on KEX algorithms) within C code, without using the compiled OpenSSL executable. Are there any source files that can give me examples of this?

**Reply by baentsch** [reply]
Sure -- you then want to use [liboqs](https://github.com/open-quantum-safe/liboqs). Example KEM code [here](https://github.com/open-quantum-safe/liboqs/blob/main/tests/example_kem.c).

**Reply by NovakovicA** [reply]
Oh okay, thought so, but I guessed there was maybe a better way. Thank you!

**Reply by baentsch** [reply]
>  there was maybe a better way.

In which way better? That code allows you to call and run any QSC KEM via C.

**Reply by NovakovicA** [reply]
I was thinking of using the OpenSSL EVP functions with these algorithms.

**Reply by baentsch** [reply]
> I was thinking of using the OpenSSL EVP functions with these algorithms.

Why? Whatever, I'm not entirely sure that that's possible in OQS-OpenSSL(1.1.1). But it's certain to work using [oqs-provider](https://github.com/open-quantum-safe/oqs-provider) in OpenSSL3: The [OpenSSL3 TLS logic consumes all crypto functionality via the EVP API and any provider](https://www.openssl.org/docs/OpenSSLStrategicArchitecture.html), incl. oqs-provider, is plugged underneath that interface.
